web/src/nav/StatusMessagesNavItem.tsx

```diff
@@ -0,0 +1,129 @@
+    private subscriptions = new Subscription()
+
+    private toggleIsOpen = () => this.setState(prevState => ({ isOpen: !prevState.isOpen }))
+    public state: State = { isOpen: false, messages: [] }
+
+    public componentDidMount(): void {
+        if (this.props.messages) {
+            this.setState({ messages: this.props.messages })
+        }
+
+        this.subscriptions.add(
+            timer(0, REFRESH_INTERVAL_MS)
+                .pipe(concatMap(fetchAllStatusMessages))
+                .subscribe(messages => this.setState({ messages }))
+        )
+    }
+
+    public componentWillUnmount(): void {
+        this.subscriptions.unsubscribe()
+    }
+
+    private renderMessage(message: GQL.IStatusMessage): JSX.Element | null {
+        switch (message.type) {
+            case GQL.StatusMessageType.CLONING:
+                return (
+                    <div key={message.message} className="status-messages-nav-item__entry">
+                        <h4 className="status-messages-nav-item__entry-title">Repositories updating</h4>
+                        <p className="status-messages-nav-item__entry-copy">{message.message}</p>
+                        {this.props.isSiteAdmin && (
+                            <p className="status-messages-nav-item__entry-link">
+                                <Link to={'/site-admin/external-services'}>Configure external services</Link>
+                            </p>
+                        )}
+                    </div>
+                )
+            default:
+                return <DropdownItem>{message.message}</DropdownItem>
+        }
+    }
+
+    public render(): JSX.Element | null {
+        const hasMessages = this.state.messages.length > 0
+        const cloning = this.state.messages.some(({ type }) => type === GQL.StatusMessageType.CLONING)
+        return (
+            <ButtonDropdown
+                isOpen={this.state.isOpen}
+                toggle={this.toggleIsOpen}
+                className="nav-link py-0 status-messages-nav-item__nav-link"
+            >
+                <DropdownToggle caret={false} className="bg-transparent d-flex align-items-center" nav={true}>
+                    {cloning ? (
+                        <CloudSyncIcon
+                            className="icon-inline"
+                            {...(!this.state.isOpen && { 'data-tooltip': 'Updating repositories...' })}
+                        />
+                    ) : (
+                        <CloudCheckIcon
+                            className="icon-inline"
+                            {...(!this.state.isOpen && { 'data-tooltip': 'Repositories up to date' })}
+                        />
+                    )}
+                </DropdownToggle>
+
+                <DropdownMenu right={true} className="status-messages-nav-item__dropdown-menu">
+                    {hasMessages ? (
+                        this.state.messages.map(this.renderMessage.bind(this))
```

Comment:
Prefer arrow functions over `bind()`
```suggestion
                        this.state.messages.map(m => this.renderMessage(m))
```

Reply:
Out of curiosity, since I grew up on `.bind()` and always saw proper usage of `bind` combined with passing-functions-as-values as a sign that someone understood `this` in JavaScript — why?

Reply:
`bind` is notoriously slow. I think there was some effort at some point to optimize it in Chrome but it shouldn't be generally assumed to be fast in all browsers. It is also not type safe in previous TypeScript versions and the type checking is pretty complex in newer versions (error messages won't be very clear). There is also a pitfall when the function takes multiple parameters (or may in the future) since `map` passes the index as the second argument and array as third.

An arrow function on the other hand is fast (has no prototype, and usually gets inlined), safe this, and is always type safe.

The only argument I have _for_ `bind` is code coverage, if the arrow function is impossible to test.

Reply:
Ah, interesting, TIL :) Thanks!